kernel: introduce `k_smp_cpu_start()` and `k_smp_cpu_resume()`

dcpleung opened the pull request:
* This promotes `z_smp_start_cpu()` to public API as `k_smp_cpu_start()` so it can be used out of tree.
* Also allow `k_smp_cpu_start()` to do additional initialization before handing off to scheduler to run threads if desired.
* Add `k_smp_cpu_resume()` to address the need to power up a CPU without re-initialize per-CPU kernel structs.
* This is a step towards eventual removal of `z_smp_thread_init()` and `z_smp_thread_swap()` exposed in `include/zephyr/kernel/internal/smp.h` specifically for SOF.

dcpleung committed:
kernel: amend wording on CONFIG_SMP_BOOT_DELAY

This extends the wording so that not only architecture code can
start secondary CPUs at a later time. Also adds a missing 'to'.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: kernel/smp_boot_delay: make test more generic

This makes the test a bit more generic so it can run on
other SMP enabled platforms. Though, we don't need CI
running this on every SMP platforms as this is not
testing the mechanism of bringing up a CPU. That is
being tested on the other SMP test. Therefore, only
enable qemu_x86_64 to be testing in CI so that this
feature will still be tested on CI.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: smp: put comment on SMP code

Adds some comments to the SMP code to, hopefully, make it
a bit more clear to future readers.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: smp_boot_delay: extend to test custom CPU start func

This extends the smp_boot_delay test to test the newly
introduced function k_smp_cpu_custom_start().

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
boards: qemu_x86_64: amend DTS to have 2 CPU nodes

qemu_x86_64 has default of 2 CPUs but the device tree only
has 1. For correctness, add another CPU node to the tree.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
modules: sof: use k_smp_cpu_custom_start for core power up

This updates the SOF to use the newly introduced
k_smp_cpu_custom_start() for secondary core power up.
This removes the need to mirror part of the SMP power up
code from Zephyr kernel into SOF tree. Also removes the need
to expose private kernel code to SOF.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: smp: remove z_smp_thread_init/_swap

This removes z_smp_thread_init() and z_smp_thread_swap() as
SOF has been updated to use k_smp_cpu_custom_start() instead.
This removes the need for SOF to mirror part of the SMP
power up code, and thus these two functions are no longer
needed.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
@kv2019i https://github.com/zephyrproject-rtos/sof/pull/32 is my crude attempt at changing the SOF secondary core power up code. I am not totally sure on the power management side. Please take a look and see if there are any issues.

cfriedt reviewed (dismissed):
LGTM

nashif commented:
@lyakh  FYI

kv2019i commented:
FYI @RanderWang who has been also looking at cleaning up this very bit of code.

peter-mitsis commented on include/zephyr/kernel/smp.h:
```diff
@@ -0,0 +1,31 @@
+#define ZEPHYR_INCLUDE_KERNEL_SMP_H_
+
+/**
+ * @brief Start one CPU.
+ *
+ * This starts (powers up/starts up) a CPU that has stopped
+ * or powered down. After the CPU is started, the scheduler
+ * will run, and the CPU will participate in execution of
+ * threads.
```
"one" -> "a"

I think this opening paragraph needs a little rework. Maybe something like the following (which is itself incomplete) ...

> This routine is used to manually start the CPU specified by _id_. It may be called to restart a CPU that had been stopped or powered down, as well as _some other scenario_. After the CPU has finished initialization, the CPU will be ready to participate in thread scheduling and execution.

dcpleung replied:
Updated the working.

peter-mitsis commented on include/zephyr/kernel/smp.h:
```diff
@@ -28,4 +30,30 @@
 void k_smp_cpu_start(int id);
 
+/**
+ * @brief Start one CPU with additional initialization callback.
```
Similar comments as those for k_smp_cpu_start()

dcpleung replied:
This function is removed so this is no longer needed.

peter-mitsis commented on kernel/smp.c:
```diff
@@ -115,10 +115,21 @@ static inline FUNC_NORETURN void smp_init_top(void *arg)
 	struct k_thread dummy_thread;
 	struct cpu_start *csd = arg;
 
+	/* Let the CPU start caller know that this CPU
+	 * has powered up and ready to be initialized.
+	 */
```
I wish I could think of a more concise phrasing than "the CPU start caller", but we can polish it slightly with "Inform xxx" instead of "Let xxx know".

Also, "and ready" -> "and is ready"

dcpleung replied:
Updated.

peter-mitsis commented on kernel/smp.c:
```diff
@@ -115,10 +115,21 @@ static inline FUNC_NORETURN void smp_init_top(void *arg)
+	 */
 	wait_for_start_signal(&csd->start_flag);
+
+	/* Initialization the dummy thread struct so that
```
Initialization -> Initialize

dcpleung replied:
Updated.

peter-mitsis commented on kernel/smp.c:
```diff
@@ -145,10 +157,21 @@ static void start_cpu(int id, smp_custom_init_fn fn, void *arg)
 	arch_start_cpu(id, z_interrupt_stacks[id], CONFIG_ISR_STACK_SIZE,
 		       smp_init_top, &cpu_start_data);
+
+	/* Wait till the newly powered up CPU to signal that
```
till -> until

dcpleung replied:
Updated.

peter-mitsis reviewed (commented):
Overall, I think this looks good--just a couple of minor nitpicks.

andyross commented on include/zephyr/kernel/smp.h:
```diff
@@ -0,0 +1,59 @@
+ * @note This unconditionally starts a CPU regardless of
+ *       its current state. Depending on the architecture,
+ *       it may have no effect or very bad side effect.
+ *       Use this with caution.
```
Nitpick, but a comment saying effectively "what this function does is going to depend on the SOC" is probably a good clue that it should still be a SOC-provided API :)

Can we tighten up the spec a little?   Maybe document this like "This function must not be called on currently-executing CPUs, the device must have reached its idle/off state.  Detection of that state in the general case must be provided by the platform layers."

dcpleung replied:
Updated.

andyross reviewed (commented):
Nitpick, but I mildly hate the use of the word "custom" here.

Since (1) the core feature provided is just a callback hook to run at the end of the init process, (2) CPU startup is not in general a performance path and (3) dealing with a previously-internal API where we're allowed to do anything:

Why not just eliminate the no-callback API entirely, have a single API named `k_smp_cpu_start()`, and make every user pass the callback?  If it's NULL, don't call it.

dcpleung committed:
kernel: smp: introduce k_smp_cpu_start

This renames z_smp_cpu_start() to k_smp_cpu_start().
This effectively promotes z_smp_cpu_start() into a public API
which allows out of tree usage. Since this is a new API,
we can afford to change it signature, where it allows
an additional initialization steps to be done before a newly
powered up CPU starts participating in scheduling threads
to run.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented on include/zephyr/kernel/smp.h:
```diff
@@ -0,0 +1,31 @@
+ *
+ * @warning Parallel calls to this function will result
+ *          in undesired behavior. Please make sure only
+ *          one call to this function at a time.
```
Added a spinlock.

dcpleung commented:
> Nitpick, but I mildly hate the use of the word "custom" here.
> 
> Since (1) the core feature provided is just a callback hook to run at the end of the init process, (2) CPU startup is not in general a performance path and (3) dealing with a previously-internal API where we're allowed to do anything:
> 
> Why not just eliminate the no-callback API entirely, have a single API named `k_smp_cpu_start()`, and make every user pass the callback? If it's NULL, don't call it.

I consolidated them into one `k_smp_cpu_start()`.

andyross reviewed (dismissed):
Looks great to me.

RanderWang reviewed (commented):
please check the usage in SOF : https://github.com/zephyrproject-rtos/sof/pull/32. I have one concern

nashif commented on include/zephyr/kernel/internal/smp.h:
```diff
@@ -6,18 +6,7 @@
-void z_smp_thread_swap(void);
-#endif
-
 void z_init_cpu(int id);
```
what about those internal APIs? I see they are still being used in SOF.

kv2019i reviewed (dismissed):
There's an open issue with https://github.com/zephyrproject-rtos/sof/pull/32#pullrequestreview-1720138691 , flagging a change request here before we can remove the internal interfaces.

dcpleung committed:
kernel: smp: introduce k_smp_cpu_resume

This provides a path to resume a previously suspended CPU.
This differs from k_smp_cpu_start() where per-CPU kernel
structs are not initialized such that execution context
can be saved during suspend and restored during resume.
Though the actual context saving and restoring are
platform specific.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Hm... added `k_smp_cpu_resume()` to allow powering up a CPU without re-initializing per-CPU kernel structs, which would allow per-CPU suspend/resume with execution context saving/restoring. May not be the ideal solution but does provide a path forward to eliminate non-public API usage on SOF regarding to SMP.

andyross commented on kernel/smp.c:
```diff
@@ -64,62 +103,128 @@ static void wait_for_start_signal(atomic_t *start_flag)
+	/* Power up the CPU */
 	arch_start_cpu(id, z_interrupt_stacks[id], CONFIG_ISR_STACK_SIZE,
-		       smp_init_top, start_flag);
+		       smp_init_top, (void *)csc);
```
Pedantry: all pointers will promote to "void *" automatically, you don't need the cast.

dcpleung replied:
Somehow my brain thought of static analysis tool warning about pointer conversion... probably it's the other way around where (void *) is cast into something without explicitly saying so...

Anyway, removed.

andyross commented on kernel/smp.c:
```diff
@@ -64,62 +103,128 @@ static void wait_for_start_signal(atomic_t *start_flag)
+
 #ifdef CONFIG_SYS_CLOCK_EXISTS
-	smp_timer_init();
+	if ((csc == NULL) || ((csc != NULL) && csc->reinit_timer)) {
```
Style: the `csc != NULL` test is needless

dcpleung replied:
Removed.

andyross commented on kernel/smp.c:
```diff
@@ -64,62 +103,128 @@ static void wait_for_start_signal(atomic_t *start_flag)
 	for (int i = 1; i < num_cpus; i++) {
-		start_cpu(i, &cpu_start_flag);
+		z_init_cpu(i);
+		start_cpu(i, NULL);
```
Passing the NULL here requires you to add tests in a bunch of spaces in smp_init_top() to handle the NULL pointer.  Since the `cpu_start_fn` struct is static and available, can't you just set it up (a NULL function pointer and reinit_timer==true) to do what you want?

dcpleung replied:
This is done so that if `k_smp_cpu_start()` is not used at all, the garbage collector can remove the function and the static struct from the final binary... to save a few bytes.

andyross commented on include/zephyr/kernel/smp.h:
```diff
@@ -0,0 +1,66 @@
+ * suspended CPU specified by @a id. This requires platform
+ * specific routines to save and restore execution context
+ * across suspending and resuming CPU. If in doubt, use
+ * @ref k_smp_cpu_start instead.
```
Still don't love trying to document kernel APIs in terms of what the platform layers may or may not be doing.   And as I read it this doesn't really require any special handling in the arch/soc layer anyway as far as usage of these APIs goes.

Suggest just saying what the kernel does, and talking about platform layers only when we have to declare requirements, e.g. "This function works like k_smp_cpu_start(), but does not re-initialize the kernel's internal tracking data for the CPU.  Therefore k_smp_cpu_start() must have previously been called for this CPU, and it must have verifiably reached an idle/off state (detection of which must be provided by the platform layers).  It may be used in cases where platform layers require e.g. that data on the interrupt or idle stack be preserved.  It otherwise behaves identically."

dcpleung replied:
This sounds a lot better. Fixed.

andyross reviewed (commented):
Some nitpicks.  No complaints about having a "resume" API; that seems clear and obvious to me.  I still think that we'll be struggling with this synchronous idle stuff for a long time, though.

dcpleung commented on kernel/smp.c:
```diff
@@ -64,62 +103,128 @@ static void wait_for_start_signal(atomic_t *start_flag)
+
+	/* We are only starting one CPU so we do not need to synchronize
+	 * across all CPUs using the start_flag. So just set it to 1.
+	 */
```
Removed.

dcpleung committed:
kernel: move z_init_cpu to private kernel headers

z_init_cpu() is a private kernel API so move it under
kernel/include.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
x86: add CODE_UNREACHABLE to z_x86_cpu_init

For some reason, unrelated code change triggered compiler
warning about this function returns even though it is
marked nonreturn. So add CODE_UNREACHABLE to silence
the warning, possibly to catch any errors.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

RanderWang commented:
@dcpleung  two changes (1) invoke_sched, do I misunderstand it ? (2) move z_dummy_thread_init down since resume doesn't need it for schedule, or FW panic.

````c
diff --git a/kernel/smp.c b/kernel/smp.c
index 7b09520c8a1..45ec956dff9 100644
--- a/kernel/smp.c
+++ b/kernel/smp.c
@@ -119,11 +119,6 @@ static inline void smp_init_top(void *arg)
         */
        wait_for_start_signal(&cpu_start_flag);
 
-       /* Initialize the dummy thread struct so that
-        * the scheduler can schedule actual threads to run.
-        */
-       z_dummy_thread_init(&dummy_thread);
-
 #ifdef CONFIG_SYS_CLOCK_EXISTS
        if ((csc == NULL) || csc->reinit_timer) {
                smp_timer_init();
@@ -135,11 +130,16 @@ static inline void smp_init_top(void *arg)
                csc->fn(csc->arg);
        }
 
-       if ((csc != NULL) && csc->invoke_sched) {
+       if ((csc != NULL) && !csc->invoke_sched) {
                /* Don't invoke scheduler. */
                return;
        }
 
+       /* Initialize the dummy thread struct so that
+        * the scheduler can schedule actual threads to run.
+        */
+       z_dummy_thread_init(&dummy_thread);
+
        /* Let scheduler decide what thread to run next. */
        z_swap_unlocked();
 
@@ -171,6 +171,7 @@ void k_smp_cpu_start(int id, smp_init_fn fn, void *arg)
 
        cpu_start_fn.fn = fn;
        cpu_start_fn.arg = arg;
+       cpu_start_fn.invoke_sched = true;
 
 #ifdef CONFIG_SYS_CLOCK_EXISTS
        cpu_start_fn.reinit_timer = true;
````

dcpleung commented:
Thank you. Updated.

RanderWang reviewed (approved):
LGTM. Validated on tigerlake & meterlake.

RanderWang committed:
soc: intel_adsp: call framework callback function for restore

When exiting power gated state, call the CPU start function
passed to arch_start_cpu().

Signed-off-by: Rander Wang <rander.wang@intel.com>
Signed-off-by: Daniel Leung <daniel.leung@intel.com>

tmleman reviewed (dismissed):
The changes look good but don't work. I have cherry-pick them locally and found regression on meteorlake. When enabling all cores in stress test, power-up operations fails in second iteration.

RanderWang commented:
> The changes look good but don't work. I have cherry-pick them locally and found regression on meteorlake. When enabling all cores in stress test, power-up operations fails in second iteration.

I also did stress test with all core enabled. you need to apply 3 PRs: https://github.com/zephyrproject-rtos/sof/pull/32 (apply this patch
````c
diff --git a/zephyr/lib/cpu.c b/zephyr/lib/cpu.c
index 768dc5192..86a6169a6 100644
--- a/zephyr/lib/cpu.c
+++ b/zephyr/lib/cpu.c
@@ -28,12 +28,8 @@
 extern K_KERNEL_STACK_ARRAY_DEFINE(z_interrupt_stacks, CONFIG_MP_MAX_NUM_CPUS,
 				   CONFIG_ISR_STACK_SIZE);
 
-static atomic_t ready_flag;
-
 static void secondary_init(void *arg)
 {
-	atomic_set(&ready_flag, 1);
-
 	secondary_core_init(sof_get());
 }
 
@@ -86,13 +82,8 @@ void cpu_notify_state_exit(enum pm_state state)
 {
 	if (state == PM_STATE_SOFT_OFF)	{
 #if CONFIG_MULTICORE
-		if (!cpu_is_primary(arch_proc_id())) {
-			/* Notifying primary core that secondary core successfully exit the D3
-			 * state and is back in the Idle thread.
-			 */
-			atomic_set(&ready_flag, 1);
+		if (!cpu_is_primary(arch_proc_id()))
 			return;
-		}
 #endif
 
 #ifdef CONFIG_ADSP_IMR_CONTEXT_SAVE
@@ -132,16 +123,11 @@ int cpu_enable_core(int id)
 		cpu_resume = false;
 #endif
 
-	atomic_clear(&ready_flag);
-

 
-	while (!atomic_get(&ready_flag))
-		k_busy_wait(100);
-
 	return 0;
 }
 
```` 

 https://github.com/zephyrproject-rtos/zephyr/pull/65394 and this one

dcpleung commented:
@RanderWang do you mind if I pull #65394 into this one? To make it easier to test.

RanderWang commented:
> @RanderWang do you mind if I pull #65394 into this one? To make it easier to test.

NP, please integrate it.

dcpleung commented:
Done. Pulled in the patch from #65394.

@tmleman do you mind testing it again?

nashif commented:
@tmleman do you mind testing it again?

RanderWang commented:
Let check https://github.com/thesofproject/sof/pull/8627 in SOF which include all the patches.

andyross reviewed (dismissed):
refresh +1

RanderWang commented:
CI caught some errors,  confirmed and debugging it.

RanderWang commented:
@dcpleung   update https://github.com/zephyrproject-rtos/zephyr/pull/65394 to fix SOF CI issue. please sync my PR. Thanks!

cfriedt reviewed (dismissed):
refresh+1

dcpleung commented:
The CI was `tests/net/lib/lwm2m/lwm2m_engine/net.lwm2m.lwm2m_engine` on `qemu_x86` which is unrelated to this PR.

RanderWang commented:
@kv2019i this PR was merged, so we need another PR in sof.  I will submit it if the https://github.com/thesofproject/sof/pull/8747 is merged

kv2019i commented:
@RanderWang please prepare a SOF PR that has your changes and updates the Zephyr version as well. 

We have a SOF regression with Zephyr unfortunately now so #8747 might never be merged and we may need to jump to a newer Zephyr (if a fix is found on Zephyr side) and that will then require the k_smp changes, so better to prepare that PR now (even if some SOF CI fails are expected for it).

FYI @lyakh @marc-hb @abonislawski

lyakh commented on kernel/smp.c:
```diff
@@ -80,35 +98,48 @@ void z_smp_thread_swap(void)
 
+	/* Do additional initialization steps if needed. */
+	if ((csc != NULL) && (csc->fn != NULL)) {
+		csc->fn(csc->arg);
```
@dcpleung isn't this racy in general? On all CPUs `csc` points to `cpu_start_fn`, right? And the CPU gets here after signalling the initiator CPU, the one that started this one. So the initiator CPU can go ahead and start yet another CPU potentially with different parameters. So far that shouldn't be a problem for SOF, because there `.fn` is the same for all and `.arg` is always `NULL`, but in general this seems wrong?

dcpleung replied:
Yes, there is potential for conflicts here. So the caller must take precautions to avoid that situation. Or we can implement a per-CPU data struct (but this is going to take up some space).

marc-hb commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -180,6 +181,11 @@ void power_gate_exit(void)
+	/* Secondary core is resumed by set_dx */
+	if (arch_proc_id()) {
+		mp_resume_entry();
+	}
```
This has been causing regressions and test failures (https://github.com/thesofproject/sof/issues/8818). Tentative removal of this part in:
- #68366
